projects/slap_manipulation/src/slap_manipulation/policy/peract_utils.py

```diff
@@ -0,0 +1,324 @@
+    geometry = trimesh_voxel_grid.as_boxes(colors=rgb)
+    scene = trimesh.Scene()
+    scene.add_geometry(geometry)
+    if show_bb:
+        assert d == h == w
+        _create_bounding_box(scene, voxel_size, d)
+    return scene
+
+
+def visualise_voxel(voxel_grid: np.ndarray,
+                    q_attention: np.ndarray = None,
+                    highlight_coordinate: np.ndarray = None,
+                    highlight_gt_coordinate: np.ndarray = None,
+                    highlight_alpha: float = 1.0,
+                    rotation_amount: float = 0.0,
+                    show: bool = False,
+                    voxel_size: float = 0.1,
+                    offscreen_renderer: pyrender.OffscreenRenderer = None,
+                    show_bb: bool = False,
+                    alpha: float = 0.5):
+    scene = create_voxel_scene(
+        voxel_grid, q_attention, highlight_coordinate, highlight_gt_coordinate,
+        highlight_alpha, voxel_size,
+        show_bb, alpha)
+    if show:
+        scene.show()
+    else:
+        r = offscreen_renderer or pyrender.OffscreenRenderer(
+            viewport_width=640, viewport_height=480, point_size=1.0)
+        s = _from_trimesh_scene(
+            scene, ambient_light=[0.8, 0.8, 0.8],
+            bg_color=[1.0, 1.0, 1.0])
+        cam = pyrender.PerspectiveCamera(
+            yfov=np.pi / 4.0, aspectRatio=r.viewport_width/r.viewport_height)
+        p = _compute_initial_camera_pose(s)
+        t = Trackball(p, (r.viewport_width, r.viewport_height), s.scale, s.centroid)
+        t.rotate(rotation_amount, np.array([0.0, 0.0, 1.0]))
+        s.add(cam, pose=t.pose)
+        color, depth = r.render(s)
+        return color.copy()
+
+
+def preprocess(img, dist='transporter'):
+    """Pre-process input (subtract mean, divide by std)."""
+
+    transporter_color_mean = [0.18877631, 0.18877631, 0.18877631]
+    transporter_color_std = [0.07276466, 0.07276466, 0.07276466]
+    transporter_depth_mean = 0.00509261
+    transporter_depth_std = 0.00903967
+
+    franka_color_mean = [0.622291933, 0.628313992, 0.623031488]
+    franka_color_std = [0.168154213, 0.17626014, 0.184527364]
+    franka_depth_mean = 0.872146842
+    franka_depth_std = 0.195743116
+
+    clip_color_mean = [0.48145466, 0.4578275, 0.40821073]
+    clip_color_std = [0.26862954, 0.26130258, 0.27577711]
+
+    # choose distribution
+    if dist == 'clip':
+        color_mean = clip_color_mean
+        color_std = clip_color_std
+    elif dist == 'franka':
+        color_mean = franka_color_mean
+        color_std = franka_color_std
+    else:
+        color_mean = transporter_color_mean
+        color_std = transporter_color_std
+
+    if dist == 'franka':
+        depth_mean = franka_depth_mean
+        depth_std = franka_depth_std
+    else:
+        depth_mean = transporter_depth_mean
+        depth_std = transporter_depth_std
+
+    # convert to pytorch tensor (if required)
+    if type(img) == torch.Tensor:
```

Comment:
Did you write all this code? can you check the perAct license to make sure its ok to copy it in?

Reply:
Yeah, this is all peract helpers. I'll check license and get back to you.